File: python/invariance/resources/query.py
from __future__ import annotations

from typing import Any

from ..http_client import HttpClient
from ..types import (
    NLQueryResult, TraceQueryOpts, StructuredTraceQuery, TraceQueryResult,
    StatsResult, StatsQuery, AgentNote, WriteNoteOpts, ToolSchema, QueryScope,
)


class QueryResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def ask(
        self, question: str, scope: QueryScope | None = None
    ) -> NLQueryResult:
        return await self._http.post("/v1/query", {"question": question, "scope": scope})

    async def traces(self, opts: TraceQueryOpts) -> TraceQueryResult:
        return await self._http.post("/v1/query/traces", opts)

    async def traces_structured(self, query: StructuredTraceQuery) -> TraceQueryResult:
        return await self._http.post("/v1/query/traces/structured", query)

    async def stats(self, opts: StatsQuery | None = None) -> StatsResult:
        return await self._http.get("/v1/query/stats", params=opts)

    async def write_note(self, opts: WriteNoteOpts) -> AgentNote:
        return await self._http.post("/v1/query/notes", opts)

    async def read_note(self, key: str) -> dict[str, Any]:
        return await self._http.get(f"/v1/query/notes/{key}")

    async def tools(self) -> dict[str, Any]:
        return await self._http.get("/v1/query/tools")

File: packages/typescript/src/resources/query.ts
import type { HttpClient } from '../http.js';
import type {
  NLQueryResult, TraceQueryOpts, StructuredTraceQuery, TraceQueryResult,
  StatsResult, StatsQuery, AgentNote, WriteNoteOpts, ToolSchema, QueryScope,
} from '../types/query.js';

export class QueryResource {
  constructor(private http: HttpClient) {}

  async ask(question: string, scope?: QueryScope): Promise<NLQueryResult> {
    return this.http.post<NLQueryResult>('/v1/query', { question, scope });
  }

  async traces(opts: TraceQueryOpts): Promise<TraceQueryResult> {
    return this.http.post<TraceQueryResult>('/v1/query/traces', opts);
  }

  async tracesStructured(query: StructuredTraceQuery): Promise<TraceQueryResult> {
    return this.http.post<TraceQueryResult>('/v1/query/traces/structured', query);
  }

  async stats(opts?: StatsQuery): Promise<StatsResult> {
    return this.http.get<StatsResult>('/v1/query/stats', {
      params: opts as Record<string, string | undefined>,
    });
  }

  async writeNote(opts: WriteNoteOpts): Promise<AgentNote> {
    return this.http.post<AgentNote>('/v1/query/notes', opts);
  }

  async readNote(key: string): Promise<{ data: AgentNote | null }> {
    return this.http.get<{ data: AgentNote | null }>(`/v1/query/notes/${key}`);
  }

  async tools(): Promise<{ tools: ToolSchema[] }> {
    return this.http.get<{ tools: ToolSchema[] }>('/v1/query/tools');
  }
}
